os/include/debug.h

```diff
@@ -499,7 +499,7 @@ int get_errno(void);
 #define mllvdbg(...)
 #endif
 
-#if defined(CONFIG_MM_ASSERT_ON_FAIL) && defined(CONFIG_APP_BINARY_SEPARATION) && defined (__KERNEL__)
+#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
 #define mfdbg(format, ...) lldbg(format, ##__VA_ARGS__)
```

Comment:
`lldbg` must be used only in low level code and irq handler or assert handler. We would like to avoid using lldbg in all other places. So, how about we change this as below?

```
#define mfdbg(format, ...)            if (abort_mode || up_interrupt_context()) \
                                                             lldbg(format, ##__VA_ARGS__); \
                                                        else \
                                                             dbg(format, ##__VA_ARGS__);
```

Reply:
When CONFIG_MM_ASSERT_ON_FAIL is enabled, mfdbg can be called at allocation failure before assert. Then, is it possible to print all of logs using above?

Reply:
https://github.com/Samsung/TizenRT/blob/589e026d2f9467362b346fd490d240e9d9a420a1/os/mm/mm_heap/mm_manage_allocfail.c#L116-L118

we are setting abort_mode to true in mm_manage_allocfail, we can move the assignment to the start of the function. Then the above definition of mfdbg works.

Reply:
ok. then, it will be the same as `heapinfo_dbg`.

Reply:
But need to check the conditionals. The heapinfo_dbg has `#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)`. I don't remember the reason. When you check done, could you leave the reason why the conditionals are necessary using comment for future?

Reply:
Yes. We still need to check conditionals `#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)`.
It is because `abort_mode` variable is only defined in kernel binary. 
If it is app side binary (with binary separation) , we will always use `dbg`.

Reply:
@swasthik22 @kishore-sn I still don't understand why you remove existed conditionals. Could you let me know?

Reply:
@swasthik22 @kishore-sn Any reply for my query?
CONFIG_MM_ASSERT_ON_FAIL will call an assert in this case, but without the config, it just prints logs and goes next.
I think if it goes to an assert, it should be lldbg, but if not, it should be dbg. 
And why should it be dbg in loadable apps?

Reply:
@swasthik22 Thank you for explanation. Then why do you add flat build config? It will be the same as loadable config.

Reply:
> mfdbg always uses dbg if CONFIG_MM_ASSERT_ON_FAIL is false, which is wrong
consider the following case with CONFIG_MM_ASSERT_ON_FAIL disabled,
whenever heap corruption is followed by a board crash (ex : heap corruption by task A, followed by usage fault in task B), mfdbg uses dbg to print the heap corruption logs in assert, which results in incomplete logs.

Regarding what you said,
we use the mfdbg at two functions. 
You are right with heap corruption function, it prints corrupted heap info before assert or in assert.
But allocation failure function does not call the assert with disabled CONFIG_MM_ASSERT_ON_FAIL, then no need to check the flag.